### PR TITLE
Allow setting alternative web service url with sonar.buildbreaker.alternative.server.url property.

### PR DESCRIPTION
**Pull request by sandro-h**
Hey,

A possible solution to the issue I raised in https://github.com/SonarQubeCommunity/sonar-build-breaker/issues/16.

Let me know if that would be acceptable.

### CODE REVIEW

File: src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java [hunk 1]
```diff
@@ -110,6 +113,16 @@ private void execute() {
     checkQualityGate(wsClient, analysisId);
   }
 
+  private String getServerUrl(Properties reportTaskProps) {
+	  String altServerUrl = settings.getString(BuildBreakerPlugin.ALTERNATIVE_SERVER_URL_KEY);
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java]: Does settings.getString() not have an additional parameter override that allows specifying a default? e.g `settings.getString(BuildBreakerPlugin.ALTERNATIVE_SERVER_URL_KEY, reportTaskProps.getProperty("serverUrl")`

That would allow you to remove this method entirely.

File: src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java [hunk 2]
```diff
@@ -19,8 +19,11 @@
  */
 package org.sonar.plugins.buildbreaker;
 
+import static com.google.common.base.Strings.isNullOrEmpty;
+
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.io.Files;
+
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java]: Please do not add unnecessary whitespace-only changes

File: src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java [hunk 3]
```diff
@@ -19,8 +19,11 @@
  */
 package org.sonar.plugins.buildbreaker;
 
+import static com.google.common.base.Strings.isNullOrEmpty;
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java]: Prefer normal imports if not heavily used

File: src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java [hunk 4]
```diff
@@ -110,6 +113,16 @@ private void execute() {
     checkQualityGate(wsClient, analysisId);
   }
 
+  private String getServerUrl(Properties reportTaskProps) {
+	  String altServerUrl = settings.getString(BuildBreakerPlugin.ALTERNATIVE_SERVER_URL_KEY);
+	  if (isNullOrEmpty(altServerUrl)) {
+		  return reportTaskProps.getProperty("serverUrl");
+	  } else {
+		  LOGGER.debug("Using alternative server url ({}): {}", BuildBreakerPlugin.ALTERNATIVE_SERVER_URL_KEY, altServerUrl);
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java]: Make sure lines don't exceed 120 characters

File: src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java [hunk 5]
```diff
@@ -110,6 +113,16 @@ private void execute() {
...
+		  return altServerUrl;
+	  }
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/QualityGateBreaker.java]: Please keep indentation consistent with the rest of the file.  Use indentation of 2 spaces.  Do not use tabs.

File: src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java [hunk 6]
```diff
@@ -58,6 +58,11 @@
     name = "Forbidden configuration parameters",
     description = "Comma-separated list of <code>key=value</code> pairs that should break the build.",
     global = true,
+    project = false),
+  @Property(key = BuildBreakerPlugin.ALTERNATIVE_SERVER_URL_KEY,
+    name = "Alternative server url",
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java]: url -> URL

File: src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java [hunk 7]
```diff
@@ -58,6 +58,11 @@
...
+    description = "Url to use for REST requests instead of the default sonar.core.serverBaseURL",
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java]: Please change the description string to:

`"URL to use for web service requests.  If unset, uses the <code>serverUrl</code> property from <code>${sonar.working.directory}/report-task.txt</code>."`

Use string concatenation to break this long string, ensuring lines do not exceed 120 characters.

File: src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java [hunk 8]
```diff
@@ -75,6 +80,8 @@
 
   public static final String FORBIDDEN_CONF_KEY = "sonar.buildbreaker.forbiddenConf";
 
+  public static final String ALTERNATIVE_SERVER_URL_KEY = "sonar.buildbreaker.alternative.server.url";
```
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java]: Please change the key name to `sonar.buildbreaker.alternativeServerUrl`

File: src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java (hunk 7 above)
Comment (mjdetullio) [inline src/main/java/org/sonar/plugins/buildbreaker/BuildBreakerPlugin.java]: Please also add this property with matching description to the configuration properties table in `README.md`

**Comment by mjdetullio** [comment]
Also, please use a maximum of 72 characters per line for your commit message.  Use a summary line, then a blank line, then details.  Reference issue # if possible  e.g.

```
Fix #16: Allow setting alternative web service URL via property.

A configurable server URL will allow the plugin to bypass a (reverse)
proxy in the user's SonarQube server configuration.
```

Ensure your PR is squashed into a single commit.

**Comment by sandro-h** [comment]
Thanks for all the feedback. I made the changes you suggest, however before I attempt to squash commits, I wanted to ask what the effect of that is to this pull request and whether it matters. From my understand squashing and (force) pushing the single commit will cause the commit history to get rewritten. Which is fine for the fork since it's not used, but I have no idea if it will mess with the pull request.

**Comment by mjdetullio** [comment]
When you push the rewritten history up to your fork, the PR will automatically be updated.  No harm done.

**Comment by sandro-h** [comment]
I committed a squashed version with the proposed changes, but forgot to comment it here explicitly. Not sure if you got notified of the update.
